src/agent/onefuzz-task/src/tasks/utils.rs

```diff
@@ -96,3 +96,157 @@ pub fn parse_key_value(value: String) -> Result<(String, String)> {
+        if exists(&resolved).await {
+            return Ok(Some(resolved));
+        }
+    }
+
+    // Case: legacy `subpath`.
+    //
+    // The `setup_dir`-relativized `subpath` was prefixed server-side with the hardcoded
+    // relative path `setup`. We only expect to see this after upgrading deployments that
+    // have pending tasks with legacy configs.
+    {
+        let subpath = subpath.strip_prefix("setup").unwrap_or(subpath);
+        let resolved = setup_dir.join(subpath);
+        if exists(&resolved).await {
+            return Ok(Some(resolved));
+        }
+    }
+
+    Ok(None)
+}
+
+async fn exists(path: impl AsRef<Path>) -> bool {
+    fs::metadata(path).await.is_ok()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::Path;
+
+    use anyhow::Result;
+    use tempfile::TempDir;
+
+    use super::*;
+
+    fn init_setup_dir(actual: impl AsRef<Path>) -> Result<TempDir> {
+        let actual = actual.as_ref();
+
+        let setup_dir = TempDir::new()?;
+
+        // If the actual file is nested in any subdirectories of `setup`, ensure that the
+        // intermediates exist.
+        if let Some(parent) = actual.parent() {
+            let intermediate = setup_dir.path().join(parent);
+            std::fs::create_dir_all(intermediate)?;
+        }
+
+        // Create an (empty) file on-disk for the file being referenced.
+        std::fs::write(setup_dir.path().join(actual), "")?;
+
+        Ok(setup_dir)
+    }
+
+    async fn test_case(relative: impl AsRef<Path>, given: impl AsRef<Path>) -> Result<()> {
+        let relative = relative.as_ref();
+
+        let setup_dir = init_setup_dir(relative)?;
+
+        let expected = setup_dir.path().join(relative);
+        let resolved = resolve_setup_relative_path(setup_dir.path(), given).await?;
+
+        assert_eq!(Some(expected), resolved);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_resolve_setup_relative_path_root() -> Result<()> {
+        const RELATIVE: &str = "fuzz.exe";
+
+        test_case(RELATIVE, "fuzz.exe").await?;
+        test_case(RELATIVE, "setup/fuzz.exe").await?;
+        test_case(RELATIVE, "{setup_dir}/fuzz.exe").await?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_resolve_setup_relative_path_nested() -> Result<()> {
+        const RELATIVE: &str = "x/fuzz.exe";
+
+        test_case(RELATIVE, "x/fuzz.exe").await?;
+        test_case(RELATIVE, "setup/x/fuzz.exe").await?;
+        test_case(RELATIVE, "{setup_dir}/x/fuzz.exe").await?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_resolve_setup_relative_path_double_nested() -> Result<()> {
+        const RELATIVE: &str = "x/y/fuzz.exe";
+
+        test_case(RELATIVE, "x/y/fuzz.exe").await?;
+        test_case(RELATIVE, "setup/x/y/fuzz.exe").await?;
+        test_case(RELATIVE, "{setup_dir}/x/y/fuzz.exe").await?;
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_resolve_setup_relative_path_nested_setup() -> Result<()> {
```

Comment:
This tests the unlikely edge case that we have something like a _new_, non-legacy `target_exe` that is prefixed by `setup/` because the target genuinely exists in a path like `{setup_dir}/setup/fuzz.exe`.